presto-native-execution/presto_cpp/main/http/filters/JwtTokenFilter.cpp

```diff
@@ -0,0 +1,141 @@
+ * limitations under the License.
+ */
+
+#include "presto_cpp/main/http/filters/JwtTokenFilter.h"
+#include <folly/Range.h>
+#include <folly/io/IOBufQueue.h>
+#include <folly/ssl/OpenSSLHash.h>
+#include <jwt-cpp/jwt.h>
+#include <jwt-cpp/traits/kazuho-picojson/traits.h>
+#include <proxygen/httpserver/ResponseBuilder.h>
+#include "presto_cpp/main/common/Configs.h"
+#include "presto_cpp/main/http/HttpConstants.h"
+
+namespace facebook::presto::http::filters {
+
+// If the internal communication is not enabled the filter
+// is not added to process incoming requests.
+void JWTTokenFilter::onRequest(
+    std::unique_ptr<proxygen::HTTPMessage> msg) noexcept {
+  // If JWT was enabled a secret must be set and as a result this code is
+  // executed.
+  try {
+    auto token = msg->getHeaders().getSingleOrEmpty(kPrestoInternalBearer);
+    if (token.empty()) {
+      // The header did not contain the expected token.
+      // Treat as missing claim.
+      std::error_code ec{jwt::error::token_verification_error::missing_claim};
+      throw jwt::error::token_verification_exception(ec);
+    }
+
+    // Build the signature key from the secret and validate the signature.
+    auto secretHash = std::vector<uint8_t>(SHA256_DIGEST_LENGTH);
+    folly::ssl::OpenSSLHash::sha256(
+        folly::range(secretHash),
+        folly::ByteRange(folly::StringPiece(
+            SystemConfig::instance()->internalCommunicationSharedSecret())));
+
+    // Decode and verify the JWT.
+    auto decodedJwt = jwt::decode<jwt::traits::kazuho_picojson>(token);
+    auto verifier = jwt::verify<jwt::traits::kazuho_picojson>().allow_algorithm(
+        jwt::algorithm::hs256{std::string(
+            reinterpret_cast<char*>(secretHash.data()), secretHash.size())});
+    verifier.verify(decodedJwt);
+
+    auto subject = decodedJwt.get_subject();
+    // The nodeId of the requester is the subject. Check if it was set.
+    if (subject.empty()) {
+      std::error_code ec{jwt::error::token_verification_error::missing_claim};
+      throw jwt::error::token_verification_exception(ec);
+    }
+    // Passed the verification, move the message along.
+    Filter::onRequest(std::move(msg));
+  } catch (const jwt::error::token_verification_exception& e) {
+    sendUnauthorizedResponse();
+  } catch (const jwt::error::signature_verification_exception& e) {
+    sendUnauthorizedResponse();
+  } catch (const std::system_error& e) {
+    sendGenericErrorResponse();
+  }
+}
+
+void JWTTokenFilter::onBody(std::unique_ptr<folly::IOBuf> body) noexcept {
+  if (requestRejected_) {
+    return;
+  }
+  if (upstream_) {
+    // forward
+    upstream_->onBody(std::move(body));
+  }
+}
+
+void JWTTokenFilter::onEOM() noexcept {
+  if (requestRejected_) {
+    return;
+  }
+  if (upstream_) {
+    // forward
+    upstream_->onEOM();
+  }
+}
+
+void JWTTokenFilter::requestComplete() noexcept {
+  if (requestRejected_) {
+    delete this;
```

Comment:
How do we ensure there is never a `double delete`?

Reply:
I don't think this can happen as this is called only once (for a request). But I think what this does is a possible leak where the filter is not deleted in the successful request case. I have to change the handling a bit.